helita/sim/bifrost.py

```diff
@@ -329,44 +360,196 @@ def _get_simple_var_xy(self, var, order='F', mode='r'):
         return np.memmap(filename, dtype=self.dtype, order=order, offset=offset,
                          mode=mode, shape=(self.nx, self.ny))
 
-
-    def _get_composite_var(self, var):
+    def boundcut(self, name,var):
+        if ((np.shape(var))[2] != self.nz):
+            var = var[:,:,self.nb:self.nb+self.nz]
+            setattr( self, name, var)
+    def _get_composite_var(self, var, order='F'):
         """
         Gets composite variables (will load into memory).
         """
-        from . import cstagger
+        from . import cstagger as cs
         if var in ['ux', 'uy', 'uz']:  # velocities
-            rdt = self.r.dtype
-            cstagger.init_stagger(self.nzb, self.z.astype(rdt),
-                                  self.zdn.astype(rdt))
-            p = self.get_var('p' + var[1])
-            if getattr(self, 'n' + var[1]) < 5:
-                return p / self.r   # do not recentre for 2D cases (or close)
+            if not hasattr(self, 'p'+ var[1]): setattr( self, 'p' + var[1],self.get_var('p' + var[1],self.snap))
+            if not hasattr(self, 'r'): self.r = self.variables['r']= self.get_var('r',self.snap)
+            self.boundcut('r',self.r)
+            self.boundcut('p'+ var[1],getattr(self,'p'+ var[1]))
+            if (getattr(self, 'n' + var[1]) < 5):
+                return getattr(self,'p'+ var[1]) / self.r   # do not recentre for 2D cases (or close)
             else:  # will call xdn, ydn, or zdn to get r at cell faces
-                return p / getattr(cstagger, var[1] + 'dn')(self.r)
+                rdt = self.r.dtype
+                cs.init_stagger(self.nz, self.dx, self.dy, self.z.astype(rdt), self.zdn.astype(rdt), self.dzidzup.astype(rdt), self.dzidzdn.astype(rdt))
```

Comment:
Always call `init_cstagger` with `self.nzb` instead of `self.nz` to avoid boundary problems.